### PR TITLE
[WIP] testing/thrift: new aport

### PR DESCRIPTION
**Pull request by cfriedt**
Hi all. I don't expect this PR to be approved as-is, and expect that it will go through several rounds, but I thought I would get some early feedback.

Some things are obvious, e.g. I should probably add checksums of patches, plus at least get the python unit tests passing first. Will remove the README and Dockerfile for the final PR, and will squash commits.

The main reason for the delay is getaddrinfo(3) missing AI_ADDRCONFIG support in musl, which will be in the musl-1.1.20 release. So this package should not be considered until then, or unless >= git commit 187bcc3bf40 can be used (unlikely).

There are a few bugs covered in the README. Some seem to be known issues with Python, Go. There is a Java issue stemming from maven that I haven't nailed down yet. The Python and Go issues I think are very likely due to musl quirks. Should be fixable, I hope. Maybe others in the community have some suggestions.

The Dockerfile should allow you to easily reproduce the results with something like "docker build -t foo ."

In any case, what is working now is:

thrift         (the compiler)
libthrift      (C++ bindings)
libthrift-dev  (C++ bindings - development files)
libthriftc     (c_glib bindings)
libthriftc-dev (c_glib bindings - development files)
lua-thrift     (Lua bindings)
perl-thrift    (Perl bindings)
node-thrift    (Nodejs bindings, although users might as well just use "npm install thrift")

### CODE REVIEW

**Comment by tcely** [comment]
Please add `[WIP]` to the PR title.

File: community/thrift/APKBUILD [hunk 1]
```diff
@@ -0,0 +1,222 @@
+# Contributor: Christopher Friedt <chrisfriedt@gmail.com>
+# Maintainer: Christopher Friedt <chrisfriedt@gmail.com>
+pkgname=thrift
+pkgver=0.11.0
+pkgrel=0
+pkgdesc="Thrift is a lightweight, language-independent software stack with an associated code generation mechanism for RPC"
+url="https://github.com/apache/${pkgname}/archive/${THRIFT_VER}.tar.gz"
```
Comment (tcely) [inline community/thrift/APKBUILD]: `url` is intended more as a home page than a link to the source.

File: community/thrift/APKBUILD [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+#    cd "$builddir"
+#    pkgdesc="${pkgname} ruby bindings"
+#    depends="ruby"
+#}
+
+#rust() {
+#    cd "$builddir"
+#    pkgdesc="${pkgname} rust bindings"
+#    depends="rust"
+#}
```
Comment (tcely) [inline community/thrift/APKBUILD]: You don't need to comment out these split functions.

File: community/thrift/APKBUILD (hunk 1 above)
Comment (cfriedt) [reply community/thrift/APKBUILD]: good point, thanks!

File: community/thrift/APKBUILD (hunk 2 above)
Comment (cfriedt) [reply community/thrift/APKBUILD]: The'yre currently unimplimented anyway.

**Comment by tcely** [comment]
The good news is musl was upgraded and python2 bindings are not needed anymore. Please work on python3 bindings instead.

**Comment by cfriedt** [comment]
> ~Please add `[WIP]` to the PR title.~
> 
> Please move this new aport to `testing` instead of `community`.

Sorry - been awfully busy. Just moved to testing.

**Comment by cfriedt** [comment]
@tcely I had to disable python3 bindings unfortunately because of an SSL issue. However, C, C++, Lua, and Perl are all working, and I managed to update it to alpine:edge and thrift-0.12.0 .

Builds are working 100% reproducibly using the Docker environment I included.

My next steps will likely be
1. support python3 again and disable python2 (potentially some creative solutions to get SSL working)
1. build a java package (it's still a fairly popular programming language..)
1. build a rust package (it's gaining in popularity)
1. build a ruby package

I would expect the majority of users to use C++ / Java / Python / C, so getting those bases covered is the priority.

**Comment by cfriedt** [comment]
Update APKBUILD with thrift-0.12.0 tarball checksum

**Comment by cfriedt** [comment]
Again, build works fine with the provided testing/thrift/Dockerfile . Not sure what the difference is between that and this ci loop.

**Comment by tcely** [comment]
Please rebase against current be master branch.

You also need to fix your split function.
```
>>> libthriftc*: Running split function c...
mv: can't rename '/home/travis/build/alpinelinux/aports/testing/thrift/pkg/thrift-tmp/usr/lib/libthrift_c_glib.so*': No such file or directory
>>> ERROR: libthriftc*: c failed
>>> ERROR: thrift*: prepare_subpackages failed
>>> ERROR: thrift: rootpkg failed
```

File: testing/thrift/APKBUILD [hunk 3]
```diff
@@ -0,0 +1,237 @@
+# Contributor: Christopher Friedt <chrisfriedt@gmail.com>
+# Maintainer: Christopher Friedt <chrisfriedt@gmail.com>
+pkgname=thrift
+pkgver=0.12.0
+pkgrel=0
+pkgdesc="Thrift is a lightweight, language-independent software stack with an associated code generation mechanism for RPC"
```
Comment (maxice8) [inline testing/thrift/APKBUILD]: Very long, can be reduced

File: testing/thrift/APKBUILD [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+_luaver="5.1"
+
+# XXX: @CF: 20180612: add mono to depends and mono-dev to depends_dev after alpine is > 3.8
+# XXX: @CF: 20180614: maven tries to install into /root/.m2 inside of fakeroot, and fails. The java libraries can wait # apache-ant maven openjdk8
+# XXX: @CF: 20180711: go fails due to this https://github.com/golang/go/issues/14481 (glibc-ism)
+# XXX: @CF: 20180718: make check fails in test/py with "[Errno 98] Address in use"
+# XXX: @CF: 20180718: nodejs builds fine, but users can more easily install it with "npm install thrift" # nodejs 
+# XXX: @CF: 20190508: python support is broken on alpine:edge
+depends=""
```
Comment (maxice8) [inline testing/thrift/APKBUILD]: empty variables can be removed for brevity

File: testing/thrift/APKBUILD [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+depends_dev=""
+#makedepends="boost-dev libevent-dev openssl-dev zlib-dev boost libevent openssl ruby rust zlib apache-ant autoconf automake bison file flex libtool lua${_luaver} lua${_luaver}-dev pkgconfig py-six python2 python3 python2-dev python3-dev py-openssl py3-openssl py2-backports.ssl_match_hostname py2-six"
```
Comment (maxice8) [inline testing/thrift/APKBUILD]: remove

File: testing/thrift/APKBUILD [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+makedepends="boost-dev libevent-dev openssl-dev zlib-dev boost libevent openssl ruby rust zlib apache-ant autoconf automake bison file flex libtool lua${_luaver} lua${_luaver}-dev pkgconfig"
```
Comment (maxice8) [inline testing/thrift/APKBUILD]: cut at 80 chars or below and remove `boost` `libevent` `openssl` `zlib`

File: testing/thrift/APKBUILD [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+install=""
+subpackages="
+    lib${pkgname}c:c
+    lib${pkgname}c-dev:cdev
+    lib${pkgname}:cpp
+    lib${pkgname}-dev:cppdev
+    lua-${pkgname}:lua
+    perl-${pkgname}:perl
```
Comment (maxice8) [inline testing/thrift/APKBUILD]: add :noarch

File: testing/thrift/APKBUILD [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+depends_dev=""
+#makedepends="boost-dev libevent-dev openssl-dev zlib-dev boost libevent openssl ruby rust zlib apache-ant autoconf automake bison file flex libtool lua${_luaver} lua${_luaver}-dev pkgconfig py-six python2 python3 python2-dev python3-dev py-openssl py3-openssl py2-backports.ssl_match_hostname py2-six"
+makedepends="boost-dev libevent-dev openssl-dev zlib-dev boost libevent openssl ruby rust zlib apache-ant autoconf automake bison file flex libtool lua${_luaver} lua${_luaver}-dev pkgconfig"
+install=""
+subpackages="
+    lib${pkgname}c:c
+    lib${pkgname}c-dev:cdev
+    lib${pkgname}:cpp
+    lib${pkgname}-dev:cppdev
+    lua-${pkgname}:lua
```
Comment (maxice8) [inline testing/thrift/APKBUILD]: add :noarch

File: testing/thrift/APKBUILD [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+
+check() {
+    cd "$builddir"
+    make check
+}
+
+package() {
+    local tpkgdir="${pkgdir}-tmp"
+    cd "$builddir"
+    make DESTDIR="$tpkgdir" install
```
Comment (maxice8) [inline testing/thrift/APKBUILD]: you can install to $pkgdir and then move from there, no need to do this dance around a mythical -tmp package dir

File: testing/thrift/APKBUILD [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+    install -d "$pkgdir"/usr/bin
+    install "$tpkgdir"/usr/bin/thrift "$pkgdir"/usr/bin/thrift
+}
+
+c() {
+    local tpkgdir="${pkgdir}-tmp"
+    pkgdesc="${pkgname} c bindings"
+    depends="glib openssl zlib"
```
Comment (maxice8) [inline testing/thrift/APKBUILD]: should be found automatically via the packag emanager

File: testing/thrift/abuild.conf [hunk 11]
```diff
@@ -0,0 +1,29 @@
+export CFLAGS="-Os -fomit-frame-pointer"
```
Comment (maxice8) [inline testing/thrift/abuild.conf]: Why is this being added ?

File: testing/thrift/README [hunk 12]
```diff
@@ -0,0 +1,82 @@
+As soon as musl-1.1.20 is released, the thrift family of packages (and sub-packages) is
```
Comment (maxice8) [inline testing/thrift/README]: Shouldn't this be installed from upstream ?

File: testing/thrift/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,72 @@
+# XXX: @CF: 20190405: Switch to alpine:edge soon, as it uses musl-1.1.22 by default
```
Comment (maxice8) [inline testing/thrift/Dockerfile]: Why vendoring a dockerfile ?

**Pull request closed without merging**